File: donuts.py
import numpy as np
from numpy import sin, cos
import os


def RX(t) -> np.ndarray:
    return np.array([
        [      1,       0,       0],
        [      0,  cos(t), -sin(t)],
        [      0,  sin(t),  cos(t)] 
    ])

def RY(t) -> np.ndarray:
    return np.array([
        [  cos(t),      0, sin(t)],
        [       0,      1,      0],
        [ -sin(t),      0, cos(t)] 
    ])

def RZ(t) -> np.ndarray:
    return np.array([
        [  cos(t), -sin(t),     0],
        [  sin(t),  cos(t),     0],
        [       0,       0,     1] 
    ])


def render_frame(A, B) -> None:

    screen_width, screen_height = 20, 40
    theta_spacing, phi_spacing = 0.07, 0.02

    R1, R2 = 1, 2
    K2 = 5
    K1 = screen_width*K2*3/(8*(R1+R2))

    RAY = np.array([0, 1, -1])

    output = np.array([[' ']*(screen_height+1)]*(screen_width+1))
    zbuffer = np.array([[0]*(screen_height+1)]*(screen_width+1))

    for phi in np.arange(0, 2*np.pi, phi_spacing):
        for theta in np.arange(0, 2*np.pi, theta_spacing):

            RotMat = RY(phi)@RX(A)@RZ(B)

            # ドーナツの断面
            donuts2d = np.array([R2+R1*cos(theta), R1*sin(theta), 0])

            # 断面を回転させて立体のドーナツに
            donuts3d = donuts2d@RotMat
            x, y, z = donuts3d
            screen_x , screen_y = int(screen_width/2 + (K1*x)/(K2+z)), int(screen_height/2 + (K1*y)/(K2+z))

            # 光の計算
            N = np.array([cos(theta), sin(theta), 0])@RotMat
            L = np.sum(N*RAY)

            # 光が届く
            if L > 0:
                # 今までで近い場合
                if z < zbuffer[screen_x][screen_y]:

                    # z-bufferとscreenの更新
                    output[screen_x][screen_y] = ".,-~:;=!*#$@"[int(min(L*8, 11))]
                    zbuffer[screen_x][screen_y] = z
            
    os.system('clear' if os.name == 'posix' else 'cls')
    for iy in output:
        for ix in iy:
            print(ix, end='')
        print()




if __name__ == '__main__':
    A, B = 0, 0
    while True:
        render_frame(A, B)
        A += 0.2
        B += 0.1

File: donuts.cpp
#include <iostream>
#include <cmath>
#include <vector>

using namespace std;

#define SCREEN_H 24
#define SCREEN_W 24

const char luminance[12] = {'.', ',', '-', '~', ':', ';', '=', '!', '*', '#', '$', '@'};

const float theta_spacing = 0.07;
const float phi_spacing   = 0.02;

const float R1 = 1;
const float R2 = 2;
const float K2 = 5;
const float K1 = SCREEN_W*K2*3/(8*(R1+R2));

// フレームの描画
void render_frame(float A, float B){
    
    // A, B 三角関数の事前計算で高速化
    float sinA = sin(A), cosA = cos(A);
    float sinB = sin(B), cosB = cos(B);

    // z-bufferとスクリーン
    vector<vector<char>> output(SCREEN_H+1, vector<char>(2*SCREEN_W+1, ' '));
    vector<vector<float>> z_buf(SCREEN_H+1, vector<float>(2*SCREEN_W+1, 0));

    // 断面の中での回転
    for (float theta = 0; theta < 2*M_PI; theta+=theta_spacing) {

        // 事前計算で高速化
        float costheta = cos(theta), sintheta = sin(theta);

        // ドーナツの円環状に回転
        for(float phi=0; phi < 2*M_PI; phi+=phi_spacing) {

            // 事前計算で高速化
            float cosphi = cos(phi), sinphi = sin(phi);

            // 断面の作成
            float circlex = R2 + R1*costheta;
            float circley = R1*sintheta;

            // 作成した断面をドーナツに沿って回転し、空間にプロット
            float x = circlex*(cosB*cosphi + sinA*sinB*sinphi) - circley*cosA*sinB; 
            float y = circlex*(sinB*cosphi - sinA*cosB*sinphi) + circley*cosA*cosB;
            float z = K2 + cosA*circlex*sinphi + circley*sinA;
            float ooz = 1/z;  // "one over z"

            // 二次元に写像
            int xp = (int) (2*(SCREEN_W/2 + K1*ooz*x));
            int yp = (int) (SCREEN_H/2 - K1*ooz*y);

            // 光の計算
            float L = cosphi*costheta*sinB - cosA*costheta*sinphi - sinA*sintheta + cosB*(cosA*sintheta - costheta*sinA*sinphi);

            // 光が指していて、今までで一番近い場合は更新
            if (L > 0) {
                if(ooz > z_buf[yp][xp]) {
                    z_buf[yp][xp] = ooz;
                    int luminance_index = L*8;
                    output[yp][xp] = luminance[luminance_index];
                }
            }

        }

    }
    
    // フレームを書き出し
    printf("\x1b[H\n");
    for(auto line : output){
        for (int _ = 0; _ < 18; _++)printf(" ");  // pandding
        for(char ch : line){
            printf("%c", ch);
        }
        printf("\n");
    }
}

int main(int argc, char const *argv[]){

    float A = 0, B = 0;

    printf("\x1b[2J");

    while (true){
        render_frame(A, B);
        A += 0.004;
        B += 0.002;
    }
    
    return 0;
}